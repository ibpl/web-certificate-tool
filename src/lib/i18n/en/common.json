{
	"changeLanguage": "Change language",
	"changeThemeMode": "Change theme mode",
	"confirmPageLeaving": "Confirm Page Leaving",
	"details": "Details",
	"error": "Error",
	"errorFetch": "Error communicating with server.",
	"no": "No",
	"ok": "OK",
	"operation": "Operation",
	"pageLeaveConfirmation": "The changes that you made won't be saved! Are you sure that you want to leave the current page?",
	"pageNotFound": "Requested page does not exist.",
	"themeModeDark": "Dark",
	"themeModeLight": "Light",
	"themeModeSystem": "System",
	"unknownError": "Unknown error.",
	"url": "URL",
	"webCertificateTool": "Web Certificate Tool",
	"yes": "Yes"
}
