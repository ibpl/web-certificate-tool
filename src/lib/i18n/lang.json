{
	"ar": "RTL example",
	"en": "English",
	"pl": "Polski"
}
