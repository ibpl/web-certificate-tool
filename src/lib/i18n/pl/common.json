{
	"badConfig": "Nieprawidłowa zawartość pliku konfiguracyjego.",
	"badContentType": "Nieprawidłowy typ zawartości \"{contentType}\".",
	"badResponse": "Nieprawidłowa odpowiedź z serwera.",
	"changeLanguage": "Zmień język",
	"changeThemeMode": "Zmień tryb szaty",
	"confirmPageLeaving": "Potwierdź opuszczenie strony",
	"details": "Szczegóły",
	"error": "Błąd",
	"errorFetch": "Błąd komunikacji z serwerem.",
	"licenseInfo": "Web Certificate Tool jest wolnym i otwartym oprogramowaniem '<'a href=\"{url}\"'>'wydanym'<'/a'>' na warunkach licencji AGPLv3.",
	"no": "Nie",
	"ok": "OK",
	"operation": "Operacja",
	"pageLeaveConfirmation": "Wykonane przez ciebie zmiany nie zostaną zapisane! Czy na pewno chcesz opuścić bieżącą stronę?",
	"pageNotFound": "Strona o podanym adresie nie istnieje.",
	"parameterOneOf": "Parametr {parameter} musi mieć jedną z następujących wartości: {set}.",
	"requirementNotAvailable": "Niedostępne {requirement}.",
	"timeout": "Przekroczenie czasu komunikacji z serwerem.",
	"themeModeDark": "Ciemny",
	"themeModeLight": "Jasny",
	"themeModeSystem": "Systemowy",
	"unknownError": "Nieznany błąd.",
	"unsupportedBrowser": "Nieobsługiwana przeglądarka.",
	"url": "URL",
	"webCertificateTool": "Webowe narzędzie do certyfikatów",
	"yes": "Tak"
}
