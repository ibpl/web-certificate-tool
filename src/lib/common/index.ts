// SPDX-License-Identifier: AGPL-3.0-only
// SPDX-FileCopyrightText: 2024 Informatyka Boguslawski sp. z o.o. sp.k. <https://www.ib.pl>

import { loadTranslations } from '$lib/i18n';
import { settings, settingsInitialized } from '$lib/stores';
import { get } from 'svelte/store';
import { t } from '$lib/i18n';
import { convertToAppError } from '$lib/errors';
import lang from '$lib/i18n/lang.json';
import {
	PublicKeyInfo,
	PrivateKeyInfo,
	getCrypto,
	PKCS8ShroudedKeyBag,
	type ContentEncryptionAesCbcParams
} from 'pkijs';
import { fromBER } from 'asn1js';
import { Convert } from 'pvtsutils';
import FileSaver from 'file-saver';

export const CONFIG_CONTENT_TYPE = 'application/json'; // CONFIG_CONTENT_TYPE is content type of application configuration file.
export const FETCH_TIMEOUT = 10000; // FETCH_TIMEOUT is UI data fetching timeout [ms].
export const OWNER_ID_MAX_LENGTH = 300; // OWNER_ID_MAX_LENGTH owner's ID lenght limit.
export const PBKDF2_ITERATION_COUNT = 600000; // PBKDF2_ITERATION_COUNT is number of PBKDF2 iterations for key encryption.

// Settings defines application settings.
export type Settings = {
	// themeMode allows to set application theme mode; supported values:
	// 	"dark" - dark theme is used,
	// 	"light" - light theme is used,
	// 	empty string - system preference is used.
	themeMode: string;

	// locale is application locale. When empty, browser locale is used.
	locale: string;

	// ownerId is key owner's ID.
	ownerId: string;
};

// isValidConfig check wheter given parameter is valid application configuration.
export function isValidConfig(arg: unknown): boolean {
	// Falsy (https://developer.mozilla.org/en-US/docs/Glossary/Falsy)
	// or something not object type is not a valid configuration.
	if (!arg || typeof arg !== 'object') {
		return false;
	}

	// themeMode (if present) must be a string, one of: "dark", "light".
	if (
		'themeMode' in arg &&
		(typeof arg.themeMode !== 'string' || (arg.themeMode !== 'dark' && arg.themeMode !== 'light'))
	) {
		return false;
	}

	// locale field (if present) must be supported locale string.
	if ('locale' in arg && (typeof arg.locale !== 'string' || !(arg.locale in lang))) {
		return false;
	}

	// ownerId field (if present) must be non-empty string not longer than OWNER_ID_MAX_LENGTH.
	if (
		'ownerId' in arg &&
		(typeof arg.ownerId !== 'string' ||
			!(arg.ownerId.length > 0 && arg.ownerId.length <= OWNER_ID_MAX_LENGTH))
	) {
		return false;
	}

	// All checks passed so we assume it's a valid configurations.
	return true;
}

// checkBrowserCompatibility checks whether browser meets requirements and throws an error if not.
export function checkBrowserCompatibility() {
	// SubtleCrypto interface of the Web Crypto API must be available.
	if (typeof crypto === 'undefined' || !('subtle' in crypto)) {
		throw <App.Error>{
			status: 418,
			message: t.get('common.unsupportedBrowser'),
			details: t.get('common.requirementNotAvailable', {
				requirement:
					'<a href="https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto#browser_compatibility">SubtleCrypto</a>'
			})
		};
	}
	try {
		getCrypto();
		/* v8 ignore next 3 */
	} catch (e) {
		throw convertToAppError(e, 'pkijs.getCrypto', t.get('common.unsupportedBrowser'));
	}
}

// tsFetch is generic fetch wrapper for fetching typed data.
async function tsFetch<T>(url: string, options = {}): Promise<T> {
	let response: Response;

	// Fetch response with connection error handling.
	try {
		const controller = new AbortController();
		const responsePromise = fetch(url, { ...options, signal: controller.signal });
		const timeout = window.setTimeout(() => controller.abort(), FETCH_TIMEOUT);
		response = await responsePromise;
		clearTimeout(timeout);
	} catch (e) {
		throw convertToAppError(
			e,
			'tsFetch',
			/* v8 ignore next 1 */
			t.get(e instanceof Error && e.name === 'AbortError' ? 'common.timeout' : 'common.errorFetch')
		);
	}

	// Throw error if response status is not 2xx.
	if (!response.ok) {
		throw <App.Error>{
			status: response.status,
			message: t.get('common.errorFetch'),
			url: url,
			operation: 'tsFetch',
			details: response.statusText
		};
	}

	// Throw error if response content type is different than CONFIG_CONTENT_TYPE.
	const contentType = response.headers.get('Content-Type');
	if (contentType !== CONFIG_CONTENT_TYPE) {
		throw <App.Error>{
			status: 418,
			message: t.get('common.badResponse'),
			url: url,
			operation: 'tsFetch',
			details: t.get('common.badContentType', { contentType: contentType })
		};
	}

	// Throw error if response is not valid JSON.
	try {
		response = await response.json();
	} catch (e) {
		throw <App.Error>{
			status: 418,
			message: t.get('common.badResponse'),
			url: url,
			operation: 'tsFetch',
			/* v8 ignore next 1 */
			details: e instanceof Error ? e.message : (e as string)
		};
	}

	return response as T;
}

// initializeEnvironmentInternalPaths is used for serialization of
// initializeEnvironmentInternal executions, separate for distinct path (map key)
// (initializeEnvironmentInternal with given path will be called only once).
const initializeEnvironmentInternalPaths = new Map();

// initializeEnvironmentInternal initializes application environemnt for given URL.
async function initializeEnvironmentInternal(url: URL) {
	try {
		// Load translations based on browser locale for t.get()
		// work properly for error messages that may be generated
		// during other initializations below.
		await loadTranslations(window.navigator.language, url.pathname);
		/* v8 ignore next 4 */
	} catch (e) {
		// Throw error to generate +error.svelte error page.
		throw convertToAppError(e, 'loadTranslations', t.get('common.errorFetch'));
	}

	const applicationSettingsInitialized = get(settingsInitialized);
	const applicationSettings = get(settings);

	// Initialize application settings only if not already initialized.
	if (!applicationSettingsInitialized) {
		// Try to fetch settings from /config.json and use it only if fetched and validated successfully.
		const configFileUrl: string = location.protocol + '//' + location.host + '/config.json';
		try {
			const response = await tsFetch<Settings>(configFileUrl, {});
			if (!isValidConfig(response)) {
				throw <App.Error>{
					status: 418,
					message: t.get('common.badConfig'),
					url: configFileUrl,
					operation: 'initializeEnvironmentInternal'
				};
			}

			// Use theme mode from config file only if defined there.
			if (response.themeMode) {
				applicationSettings.themeMode = response.themeMode;
			}

			// Use locale from config file only if defined there.
			if (response.locale) {
				applicationSettings.locale = response.locale;
			}

			// Use owner's ID config file only if defined there.
			if (response.ownerId) {
				applicationSettings.ownerId = response.ownerId;
			}
		} catch (e) {
			// Ignore if config.json does not exist (HTTP 404). Display other errors.
			const err = convertToAppError(
				e,
				'initializeEnvironmentInternal',
				t.get('common.unknownError')
			);
			if (err.status !== 404) {
				throw err;
			}
		}

		// Use theme mode defined with tm query parameter if present.
		const queryThemeMode = url.searchParams.get('tm');
		if (queryThemeMode !== null) {
			if (queryThemeMode == 'dark' || queryThemeMode == 'light') {
				applicationSettings.themeMode = queryThemeMode;
			} else {
				throw <App.Error>{
					status: 400,
					message: t.get('common.parameterOneOf', {
						parameter: 'tm',
						set: 'light, dark'
					}),
					url: url.toString()
				};
			}
		}

		// Use locale defined in l query parameter if present.
		const queryLocale = url.searchParams.get('l');
		if (queryLocale !== null) {
			const supportedLocales = Object.keys(lang);
			if (!supportedLocales.includes(queryLocale)) {
				throw <App.Error>{
					status: 400,
					message: t.get('common.parameterOneOf', {
						parameter: 'l',
						set: supportedLocales.join(', ')
					}),
					url: url.toString()
				};
			}
			applicationSettings.locale = queryLocale;
		}

		// Use owner ID defined in oid query parameter if present.
		const queryOwnerId = url.searchParams.get('oid');
		if (queryOwnerId !== null) {
			if (!(queryOwnerId.length > 0 && queryOwnerId.length <= OWNER_ID_MAX_LENGTH)) {
				throw <App.Error>{
					status: 400,
					message: t.get('common.parameterLengthMustBeInRange', {
						parameter: 'oid',
						from: 0,
						to: OWNER_ID_MAX_LENGTH
					}),
					url: url.toString()
				};
			}
			applicationSettings.ownerId = queryOwnerId;
		}

		// Save initialized settings in store.
		settings.set(applicationSettings);
		settingsInitialized.set(true);
	}

	// Load translations based on locale from application settings if not empty;
	// otherwise use browser's locale.
	await loadTranslations(
		applicationSettings.locale ? applicationSettings.locale : window.navigator.language,
		url.pathname
	);
}

// initializeEnvironment initializes application environment for given URL with
// protection against executing initialization in parallel.
//
// Note: initializeEnvironment must be executed on top of load function
// in every +page.ts and +layout.ts for environment to be ready for that
// load function /i.e. t.get() may not work without it/. Errors thrown
// by this function are ignored in load() in root +layout.ts but should
// not be ignored in other places.
export async function initializeEnvironment(url: URL) {
	// Start initializeEnvironmentInternal only if was not started already by another
	// async function for the same path.
	if (initializeEnvironmentInternalPaths.get(url.pathname) === undefined) {
		initializeEnvironmentInternalPaths.set(url.pathname, initializeEnvironmentInternal(url));
	}

	return initializeEnvironmentInternalPaths.get(url.pathname);
}

// resetEnvironment resets application environment.
export async function resetEnvironment() {
	initializeEnvironmentInternalPaths.clear();
	settingsInitialized.set(false);
}

// bufferToHex convert buffer to lowercase hex string separated with ":".
export function bufferToHex(buffer: ArrayBuffer) {
	return [...new Uint8Array(buffer)].map((b) => b.toString(16).padStart(2, '0')).join(':');
}

// formatPEM formats string to have each line length equal to 64 octets.
export function formatPEM(pemString: string): string {
	// Return unmodified string if not longer than 64 chars.
	if (pemString.length <= 64) {
		return pemString;
	}
	let pemStringFormatted = '',
		wrapIndex = 0;
	for (let i = 64; i < pemString.length; i += 64) {
		pemStringFormatted += pemString.substring(wrapIndex, i) + '\n';
		wrapIndex = i;
	}
	pemStringFormatted += pemString.substring(wrapIndex, pemString.length);
	return pemStringFormatted;
}

// toPEM creates PEM content from buffer and label.
export function toPEM(buffer: BufferSource, label: string): string {
	return [
		`-----BEGIN ${label}-----`,
		formatPEM(Convert.ToBase64(buffer)),
		`-----END ${label}-----`,
		''
	].join('\n');
}

// fromPEM creates buffer from PEM content.
export function fromPEM(pem: string): ArrayBuffer {
	const base64 = pem.replace(/-{5}(BEGIN|END) .*-{5}/gm, '').replace(/\s/gm, '');
	return Convert.FromBase64(base64);
}

// publicKeyIdentifier returns public key identifier (Subject Key Identifier) from given key pair.
export async function publicKeyIdentifier(
	publicKey: CryptoKey,
	algorithm: AlgorithmIdentifier
): Promise<string> {
	const crypto = getCrypto(true);
	const publicKeyBinary = await crypto.subtle.exportKey('spki', publicKey);
	const publicKeyInfo = new PublicKeyInfo({ schema: fromBER(publicKeyBinary).result });
	return bufferToHex(
		await crypto.digest(algorithm, publicKeyInfo.subjectPublicKey.valueBlock.valueHexView)
	);
}

// downloadFile makes browser to download file with given name, content type and content.
export function downloadFile(filename: string, contentType: string, content: ArrayBuffer) {
	const file = new File([content], filename, { type: contentType });
	FileSaver.saveAs(file);
}

// downloadPKCS8 downloads given private key in PEM formatted PKCS #8 file with optional encryption
// (when password is non-empty).
export async function downloadPKCS8(kp: CryptoKeyPair, password: string, filename: string) {
	let privateKeyBinary = await crypto.subtle.exportKey('pkcs8', kp.privateKey);
	let pemLabel = 'PRIVATE KEY';
	if (password) {
		const pkcs8Simpl = new PrivateKeyInfo({ schema: fromBER(privateKeyBinary).result });
		const pkcs8 = new PKCS8ShroudedKeyBag({ parsedValue: pkcs8Simpl });

		await pkcs8.makeInternalValues({
			password: Convert.FromUtf8String(password),
			iterationCount: PBKDF2_ITERATION_COUNT,
			hmacHashAlgorithm: 'SHA-256',
			contentEncryptionAlgorithm: <ContentEncryptionAesCbcParams>{
				name: 'AES-CBC',
				length: 256
			}
		});
		privateKeyBinary = pkcs8.toSchema().toBER(false);
		pemLabel = 'ENCRYPTED PRIVATE KEY';
	}
	downloadFile(
		filename,
		'application/x-pem-file',
		new TextEncoder().encode(toPEM(privateKeyBinary, pemLabel))
	);
}

// Callback is callback function definition.
export interface Callback {
	(): void;
}
