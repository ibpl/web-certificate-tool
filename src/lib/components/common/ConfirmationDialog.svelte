<!--
SPDX-License-Identifier: AGPL-3.0-only
SPDX-FileCopyrightText: 2024 Informatyka Boguslawski sp. z o.o. sp.k. <https://www.ib.pl>
-->

<script lang="ts">
	import { t } from '$lib/i18n';
	import Dialog, { Title, Content, Actions, InitialFocus } from '@smui/dialog';
	import Button, { Label } from '@smui/button';
	import { createEventDispatcher } from 'svelte';

	export let open: boolean; // open allows to open (when true) or close (when false) confirmation dialog.
	export let title: string; // title is string displayed in bar (cannot contain leading whitespace due to mdc-typography-baseline-top()).
	export let content: string; // content is content displayed in confirmation dialog.
	export let confirmLabel = ''; // confirmLabel is text displayed on confirm button.
	export let refuseLabel = ''; // refuseLabel is text displayed on refuse button.

	const dispatch = createEventDispatcher();
</script>

<Dialog
	bind:open
	aria-labelledby="confirmation-dialog-title"
	aria-describedby="confirmation-dialog-content"
>
	<Title id="confirmation-dialog-title">{title}</Title>
	<!-- eslint-disable-next-line svelte/no-at-html-tags -->
	<Content id="confirmation-dialog-content">{content}</Content>
	<Actions>
		<Button on:click={() => dispatch('confirm')} data-testid="button-confirmation-yes">
			<Label>{confirmLabel ? confirmLabel : $t('common.yes')}</Label>
		</Button>
		<Button
			on:click={() => dispatch('refuse')}
			defaultAction
			use={[InitialFocus]}
			data-testid="button-confirmation-no"
		>
			<Label>{refuseLabel ? refuseLabel : $t('common.no')}</Label>
		</Button>
	</Actions>
</Dialog>
