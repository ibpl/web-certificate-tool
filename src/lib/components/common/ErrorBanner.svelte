<!--
SPDX-License-Identifier: AGPL-3.0-only
SPDX-FileCopyrightText: 2024 Informatyka Boguslawski sp. z o.o. sp.k. <https://www.ib.pl>
-->

<script lang="ts">
	import { t } from '$lib/i18n';
	import Button from '@smui/button';
	import Banner, { Label, Icon, CloseReason } from '@smui/banner';
	import { mdiAlertCircle } from '@mdi/js';

	export let message: string | undefined = undefined; // message is error message displayed in error banner.

	let banner: Banner; // banner is error banner handle.
	let open = false; // open is banner opening status.
	let content: string | undefined = undefined; // content is content displayed in banners.

	// messageChangeHandler updates banner state on message changes.
	function messageChangeHandler(message: string | undefined) {
		if (message !== undefined) {
			content = message;
			if (!open) {
				open = true;
			}
			// https://github.com/hperrin/svelte-material-ui/issues/577 workaround.
			setTimeout(() => {
				banner.layout();
			}, 0);
		} else {
			if (open) {
				open = false;
			}
		}
	}
	$: messageChangeHandler(message);

	// closeHandler handles error banner close.
	function closeHandler(event: CustomEvent<{ reason: CloseReason }>) {
		switch (event.detail.reason) {
			case CloseReason.PRIMARY:
				message = undefined;
				break;
		}
	}
</script>

<Banner
	bind:this={banner}
	bind:open
	mobileStacked={true}
	on:SMUIBanner:closed={closeHandler}
	on:SMUIBanner:opened={() => {
		// https://github.com/hperrin/svelte-material-ui/issues/577 workaround.
		banner.layout();
	}}
>
	<Icon slot="icon" tag="svg" viewBox="0 0 24 24">
		<path fill="currentColor" d={mdiAlertCircle} />
	</Icon>
	<!-- eslint-disable-next-line svelte/no-at-html-tags -->
	<Label slot="label">{@html content}</Label>
	<svelte:fragment slot="actions">
		<Button>{$t('common.ok')}</Button>
	</svelte:fragment>
</Banner>
