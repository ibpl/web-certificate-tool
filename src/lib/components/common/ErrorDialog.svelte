<!--
SPDX-License-Identifier: AGPL-3.0-only
SPDX-FileCopyrightText: 2024 Informatyka Boguslawski sp. z o.o. sp.k. <https://www.ib.pl>
-->

<script lang="ts">
	import { t } from '$lib/i18n';
	import Button, { Label } from '@smui/button';
	import Dialog, { Title, Content, Actions, InitialFocus } from '@smui/dialog';
	import { errorDialogMessage } from '$lib/stores';

	let open: boolean;
	let message: string;

	// Open error dialog when erro message to be displayed is set in store.

	// errorDialogMessageChangeHandler handles errorDialogMessage value changes.
	function errorDialogMessageChangeHandler(errorDialogMessage: string | undefined) {
		if (errorDialogMessage !== undefined) {
			message = errorDialogMessage;
			open = true;
		} else {
			open = false;
		}
	}
	$: errorDialogMessageChangeHandler($errorDialogMessage);

	// closeHandler handles dialog close.
	function closeHandler() {
		// Reset error message for dialog to open on next error message set.
		$errorDialogMessage = undefined;
	}
</script>

<Dialog
	bind:open
	on:SMUIDialog:closed={closeHandler}
	aria-labelledby="error-dialog-title"
	aria-describedby="error-dialog-content"
>
	<Title id="error-dialog-title">{$t('common.error')}</Title>
	<!-- eslint-disable-next-line svelte/no-at-html-tags -->
	<Content id="error-dialog-content">{@html message}</Content>
	<Actions>
		<Button action="close" defaultAction use={[InitialFocus]}>
			<Label>{$t('common.ok')}</Label>
		</Button>
	</Actions>
</Dialog>
