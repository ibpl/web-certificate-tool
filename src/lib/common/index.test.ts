// SPDX-License-Identifier: AGPL-3.0-only
// SPDX-FileCopyrightText: 2024 Informatyka Boguslawski sp. z o.o. sp.k. <https://www.ib.pl>

import {
	checkBrowserCompatibility,
	initializeEnvironment,
	resetEnvironment,
	isValidConfig,
	FETCH_TIMEOUT,
	CONFIG_CONTENT_TYPE,
	formatPEM,
	fromPEM,
	toPEM
} from './index';
import { server } from '../../../tests/vitest-setup';
import { HttpResponse, http } from 'msw';
import lang from '$lib/i18n/lang.json';
import { OWNER_ID_MAX_LENGTH } from '$lib/common';
import { getAlgorithmParameters } from 'pkijs';

// PKI.js RSASSA-PKCS1-v1_5 parameters.
describe('getAlgorithmParameters', () => {
	test('should return expected parameters for RSASSA-PKCS1-v1_5 importKey', () => {
		expect(getAlgorithmParameters('RSASSA-PKCS1-v1_5', 'importKey')).toMatchSnapshot();
	});

	test('should return expected parameters for RSASSA-PKCS1-v1_5 generateKey', () => {
		expect(getAlgorithmParameters('RSASSA-PKCS1-v1_5', 'generateKey')).toMatchSnapshot();
	});
});

// Configuration validation.
describe('isValidConfig', () => {
	test('should recognize valid configurations', () => {
		expect(isValidConfig({ test: '' })).toBeTruthy();
		expect(isValidConfig({ themeMode: 'dark', locale: 'en', anything: 1 })).toBeTruthy();
		expect(isValidConfig({ themeMode: 'light', ownerId: '1', locale: 'pl' })).toBeTruthy();
		expect(
			isValidConfig({ themeMode: 'light', ownerId: 'a'.repeat(OWNER_ID_MAX_LENGTH), locale: 'pl' })
		).toBeTruthy();
	});

	test('should recognize invalid configurations', () => {
		expect(isValidConfig(1)).toBeFalsy();
		expect(isValidConfig({ themeMode: 1, locale: 'en' })).toBeFalsy();
		expect(isValidConfig({ themeMode: 'light', locale: 0 })).toBeFalsy();
		expect(isValidConfig({ ownerId: '' })).toBeFalsy();
		expect(isValidConfig({ ownerId: { test: 'test' } })).toBeFalsy();
		expect(isValidConfig({ ownerId: 'a'.repeat(OWNER_ID_MAX_LENGTH + 1) })).toBeFalsy();
	});
});

// Browser compatibility verification.
describe('checkBrowserCompatibility', () => {
	test('should not throw an error if crypto.subtle is available', () => {
		vi.stubGlobal('crypto', { subtle: 'dummy' });
		expect(checkBrowserCompatibility()).toBe(undefined);
	});

	test('should throw error when subtle is not available in crypto', () => {
		vi.stubGlobal('crypto', { dummy: 'test' });
		expect(() => checkBrowserCompatibility()).toThrowError('Unsupported browser.');
	});
	test('should throw error when crypto is undefined', () => {
		vi.stubGlobal('crypto', undefined);
		expect(() => checkBrowserCompatibility()).toThrowError('Unsupported browser.');
	});
});

// Environment initialization.
describe('initializeEnvironment', () => {
	beforeEach(() => resetEnvironment());

	// Fake timers required for testing timeouts without delaying test.
	beforeEach(() => {
		vi.useFakeTimers();
	});
	afterEach(() => {
		vi.restoreAllMocks();
	});

	test('should throw error on network error while fetching config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.error();
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Error communicating with server.'
		);
	});

	test(
		'should throw error on request timeout while fetching config.json',
		async () => {
			server.use(
				http.get('/config.json', async () => {
					// Simulate waiting 2x longer than configured request timeout to force timeout.
					vi.advanceTimersByTime(2 * FETCH_TIMEOUT);
					return new HttpResponse(null, { status: 404 });
				})
			);
			await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
				'Error communicating with server.'
			);
		},
		FETCH_TIMEOUT + 1000
	);

	test('should throw error on response status code 403 while fetching config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse(
					JSON.stringify({
						locale: 'pl',
						themeMode: 'light'
					}),
					{
						status: 403,
						statusText: 'Forbidden'
					}
				);
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Error communicating with server.'
		);
	});

	test(
		'should throw error on content-type x-' + CONFIG_CONTENT_TYPE + ' while fetching config.json',
		async () => {
			server.use(
				http.get('/config.json', () => {
					return new HttpResponse(
						JSON.stringify({
							locale: 'pl',
							themeMode: 'light'
						}),
						{
							headers: {
								'content-type': 'x-' + CONFIG_CONTENT_TYPE
							}
						}
					);
				})
			);
			await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
				'Bad response from server.'
			);
		}
	);

	test('should throw error on empty config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse('', {
					headers: {
						'Content-Type': CONFIG_CONTENT_TYPE
					}
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Bad response from server.'
		);
	});

	test('should throw error on invalid JSON in config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse('notjson', {
					headers: {
						'Content-Type': CONFIG_CONTENT_TYPE
					}
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Bad response from server.'
		);
	});

	test('should accept response status code 404 while fetching config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse(null, {
					status: 404,
					statusText: 'Not Found'
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).resolves.toBe(undefined);
	});

	test('should accept valid config.json without parameters', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					nothinghere: true
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).resolves.toBe(undefined);
	});

	test('should accept valid config.json with pl locale and light theme mode', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					locale: 'pl',
					themeMode: 'light',
					dummy: 123
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).resolves.toBe(undefined);
	});

	test('should accept valid config.json with en locale and dark theme mode and valid ownerId', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					themeMode: 'dark',
					locale: 'en',
					ownerId: 'test@example.com'
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).resolves.toBe(undefined);
	});

	test('should throw error on unsupported locale in config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					themeMode: 'dark',
					locale: 'EN'
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Invalid configuration file content.'
		);
	});

	test('should throw error on invalid locale value in config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					themeMode: 'dark',
					locale: { test: 1 }
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Invalid configuration file content.'
		);
	});

	test('should throw error on invalid themeMode value in config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					themeMode: 'something',
					locale: 'en'
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Invalid configuration file content.'
		);
	});

	test('should throw error on invalid onwerId value in config.json', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					themeMode: 'dark',
					ownerId: 1,
					locale: 'en'
				});
			})
		);
		await expect(initializeEnvironment(new URL(window.location.href))).rejects.toThrowError(
			'Invalid configuration file content.'
		);
	});

	test('should throw error on unsupported locale in query parameters', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse(null, { status: 404 });
			})
		);
		const url = new URL(window.location.href);
		url.searchParams.set('l', 'EN');
		await expect(initializeEnvironment(url)).rejects.toThrowError(
			'Parameter l value must be one of: ' + Object.keys(lang).join(', ') + '.'
		);
	});

	test('should throw error on unsupported tm in query parameters', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse(null, { status: 404 });
			})
		);
		const url = new URL(window.location.href);
		url.searchParams.set('tm', '2');
		await expect(initializeEnvironment(url)).rejects.toThrowError(
			'Parameter tm value must be one of: light, dark.'
		);
	});

	test('should throw error on empty oid in query parameters', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse(null, { status: 404 });
			})
		);
		const url = new URL(window.location.href);
		url.searchParams.set('oid', '');
		await expect(initializeEnvironment(url)).rejects.toThrowError(
			'Parameter oid length must be in range 0-300.'
		);
	});

	test('should throw error on too long oid in query parameters', async () => {
		server.use(
			http.get('/config.json', () => {
				return new HttpResponse(null, { status: 404 });
			})
		);
		const url = new URL(window.location.href);
		url.searchParams.set('oid', 'a'.repeat(OWNER_ID_MAX_LENGTH + 1));
		await expect(initializeEnvironment(url)).rejects.toThrowError(
			'Parameter oid length must be in range 0-300.'
		);
	});

	test('should accept valid config.json and valid query parameters', async () => {
		server.use(
			http.get('/config.json', () => {
				return HttpResponse.json({
					themeMode: 'dark',
					locale: 'en',
					ownerId: 'john1@example.com.example.com.example.com'
				});
			})
		);
		const url = new URL(window.location.href);
		url.searchParams.set('oid', 'b'.repeat(OWNER_ID_MAX_LENGTH));
		url.searchParams.set('tm', 'light');
		url.searchParams.set('l', 'pl');
		await expect(initializeEnvironment(url)).resolves.toBe(undefined);
	});
});

// Crypto tools.

describe('formatPEM', () => {
	test('should return unmodified 64-chars PEM string', () => {
		expect(formatPEM('a'.repeat(64))).toBe('a'.repeat(64));
	});
});

describe('fromPEM', () => {
	test('fromPEM(toPEM()) is identity', () => {
		expect(
			new TextDecoder().decode(fromPEM(toPEM(new TextEncoder().encode('a'.repeat(64)), 'TEST')))
		).toBe('a'.repeat(64));
	});
});
