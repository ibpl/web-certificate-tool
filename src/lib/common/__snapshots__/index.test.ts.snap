// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAlgorithmParameters > should return expected parameters for RSASSA-PKCS1-v1_5 generateKey 1`] = `
{
  "algorithm": {
    "hash": {
      "name": "SHA-256",
    },
    "modulusLength": 2048,
    "name": "RSASSA-PKCS1-v1_5",
    "publicExponent": Uint8Array [
      1,
      0,
      1,
    ],
  },
  "usages": [
    "sign",
    "verify",
  ],
}
`;

exports[`getAlgorithmParameters > should return expected parameters for RSASSA-PKCS1-v1_5 importKey 1`] = `
{
  "algorithm": {
    "hash": {
      "name": "SHA-256",
    },
    "name": "RSASSA-PKCS1-v1_5",
  },
  "usages": [
    "verify",
  ],
}
`;
